### PR TITLE
Add a new benchtool client

### PR DESCRIPTION
**Pull request by jtlisi**
- Adds a new `benchtool` tool that can be used to generate load against Prometheus remote-write and PromQL API backends.
- Add documentation for the workload configuration file of the `benchtool`

I have been iterating on this tool for a while now on this branch. However, I feel it is now stable enough to publish in a alpha form.

**note:** CI is blocked by #165 since I used some functions that are v1.16+

### CODE REVIEW

File: pkg/bench/query_runner.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+type queryRunner struct {
+	id         string
+	tenantName string
+	cfg        QueryConfig
+
+	// Do DNS client side load balancing if configured
+	dnsProvider *dns.Provider
+	remoteMtx   sync.Mutex
```
Comment (56quarters) [inline pkg/bench/query_runner.go:45]: Did this mutex used to guard a particular member variable? It's unclear to below when picking a random client what it's guarding? Everything?

File: cmd/benchtool/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,12 @@
+FROM golang:1.16.0-stretch as build
+ARG GOARCH="amd64"
+COPY . /build_dir
+WORKDIR /build_dir
+ENV GOPROXY=https://proxy.golang.org
+RUN make clean && make benchtool
+
+FROM       alpine:3.9
```
Comment (56quarters) [inline cmd/benchtool/Dockerfile:8]: 3.13 :wink:

File: pkg/bench/query_runner.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+	go q.resolveAddrsLoop(ctx)
+
+	queryChan := make(chan query, 1000)
+	for i := 0; i < 100; i++ {
+		go q.queryWorker(queryChan)
+	}
+	for _, queryReq := range q.workload.queries {
+		// every query has a ticker and a Go loop...
+		// not sure if this is a good idea but it should be fine
+		go func(q query) {
```
Comment (56quarters) [inline pkg/bench/query_runner.go:107]: Shadowing the receiver here makes it kinda hard to follow this.

File: pkg/bench/ring_check.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		for _, addr := range rs.GetAddresses() {
+			_, exists := addrMap[addr]
+			if !exists {
+				addrMap[addr] = 0
+			}
+			addrMap[addr]++
+		}
+	}
+
+	fmt.Println("ring check:")
```
Comment (56quarters) [inline pkg/bench/ring_check.go:116]: Should this use a logger as well?

File: pkg/bench/write_client.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/version"
+	"github.com/prometheus/prometheus/storage/remote"
+)
+
+const maxErrMsgLen = 512
+
+var UserAgent = fmt.Sprintf("Prometheus/%s", version.Version)
```
Comment (56quarters) [inline pkg/bench/write_client.go:27]: Shouldn't we use our own user agent here?

**Review by 56quarters — APPROVED** [review]
LGTM, few nits

File: pkg/bench/ring_check.go (hunk 4 above)
Comment (jtlisi) [reply pkg/bench/ring_check.go:116]: The reason I printed directly to stdout instead of using logs is because this is the purpose of the ring check module. So we should be able to set the log level to fatal but still get this output.

File: pkg/bench/write_client.go (hunk 5 above)
Comment (jtlisi) [reply pkg/bench/write_client.go:27]: We can, I'll update this to use benchtool

File: cmd/benchtool/Dockerfile (hunk 2 above)
Comment (jtlisi) [reply cmd/benchtool/Dockerfile:8]: 👍

File: pkg/bench/query_runner.go (hunk 1 above)
Comment (jtlisi) [reply pkg/bench/query_runner.go:45]: A better name would be addressMtx. It guards the array of addresses. There is a background process updating that slice and we want to be sure all access is mutex protected.

File: pkg/bench/query_runner.go (hunk 3 above)
Comment (jtlisi) [reply pkg/bench/query_runner.go:107]: Good point. I'll change it to `req`
